src/subprocess-posix.cc

```diff
@@ -169,11 +171,34 @@ void SubprocessSet::HandlePendingInterruption() {
     interrupted_ = SIGTERM;
 }
 
+void SubprocessSet::HandleChildDeath() {
+  if (interrupted_ != SIGCHLD)
+    return;
+  interrupted_ = false;
+  int options = WEXITED | WNOWAIT | WNOHANG;
+  for (vector<Subprocess*>::iterator i = running_.begin();
+       i != running_.end(); ) {
+    siginfo_t info = {0};
+    int res = waitid(P_PID, (id_t)(*i)->pid_, &info, options);
```

Comment:
I've changed it around a bit, so it does a loop of that sort. Now using waitpid (so no WNOWAIT), and Finish is no longer solely responsible for waiting on subprocesses.